### DISCUSSION TITLE
How do I  compile+link my code without using cmake?

### DISCUSSION THREAD

**Discussion by TassieBruce** [alandefreitas/matplotplusplus#200] [Q&A] [answered]
It would appear that the only documentation regarding compiling and linking my code with matplot++ assumes the use of cmake.  Can anyone tell me how to this for a project that does not use cmake?

It would be nice if there was a matplot++.pc file installed since enabling pkg-config would simplify things for non-cmake usage.

ETA: I'm using linux - debian 11 with matplot++ installed to /usr/local
ETA2: I use meson to build but a list of cflags and ldflags would suffice.

**Comment by alandefreitas** [comment]
If you want to compile it with another build system, then you'd have to replicate everything starting from https://github.com/alandefreitas/matplotplusplus/blob/master/CMakeLists.txt

If you just want to link a version you have installed, then you'd have to replicate the contents of the final matplot++-config.cmake file (generated from [this](https://github.com/alandefreitas/matplotplusplus/blob/master/Matplot%2B%2BConfig.cmake.in)) that's going to be in your installation directory. This matplot++-config.cmake is the cmake equivalent of a package config matplot++.pc. 

I don't think there's a way an simple way to generate a matplot++.pc from the current build scripts. Maybe there could be a small template with the flags you need. But there are still lots of dependencies to link and matplot++.pc would probably get behind matplot++-config.cmake all the time.

**Reply by TassieBruce** [reply]
Thanks for that.  I presume the important details are in
`INTERFACE_INCLUDE_DIRECTORIES` and `INTERFACE_LINK_LIBRARIES` found in 
`/usr/local/lib/cmake/Matplot++Targets.cmake`

Is there a good reason why the following does not produce useable output?

```
$ cmake --find-package -DNAME=Matplot++ -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=COMPILE
```

I get the output:

```
CMake Warning at /usr/local/lib/cmake/Matplot++/Matplot++Config.cmake:34 (message):
  This installation of Matplot++ was built with GNU.
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/CMakeFindPackageMode.cmake:183 (find_package)


CMake Error at /usr/share/cmake-3.18/Modules/CheckIncludeFileCXX.cmake:92 (try_compile):
  The TRY_COMPILE() command is not supported in --find-package mode.
Call Stack (most recent call first):
  /usr/local/lib/cmake/Matplot++/FindFilesystem.cmake:157 (check_include_file_cxx)
  /usr/share/cmake-3.18/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)
  /usr/local/lib/cmake/Matplot++/Matplot++Config.cmake:42 (find_dependency)
  /usr/share/cmake-3.18/Modules/CMakeFindPackageMode.cmake:183 (find_package)


Matplot++ not found.
CMake Error: Problem processing arguments. Aborting.
```

I suspect if this worked, meson would be able to find the dependency.

**Reply by alandefreitas** [reply]
If I'm not wrong, I think `cmake --find-package` is deprecated. In any case, it seems like it looks for matplot++ without setting a toolchain and Matplot++Config.cmake returns an error saying it was built with a different "compiler". *Maybe* removing this message "This installation of Matplot++ was built ..."  from Matplot++Config.cmake could give you what you want, but I'm really not sure.

**Reply by TassieBruce** [reply]
Yeah, I don't know what that warning is about.  I already tried building with clang and g++ and that warning and the later error are there in both cases.  Anyway, that is just a warning, the error comes later.  But `cmake --find-package` does work for other installed packages on my system like `-DNAME=Boost` and `-DNAME=PulseAudio`

Anyway, thanks for your help with this.  I have found a solution that works for me using meson build.  I'll post it as a reply to my original question outside this thread.

**Reply by alandefreitas** [reply]
> Yeah, I don't know what that warning is about

The warning is telling you are using compiler A but your matplot++ installation was built with compiler B. However, the config script is just wrong on this one. What's happening (probably) is `--find-package` simply sets no compiler toolchain. So it just compares `""` to `"GNU"` and this condition returns false. 

It's really an edge case. To avoid this in future versions, you could open a PR changing the `matplot++-config.cmake` template to only consider this condition when both strings are not empty.

**Comment by TassieBruce** [comment, marked as answer]
Answering myself with a solution that works when using the `meson` build system.

`meson` can use `cmake`'s `find_package()` function to find dependencies.  I had tried using the dependency in my `meson.build` file:
```
matplotdep = dependency('Matplot++')
```
which failed but I stupidly didn't read the error message fully.  `Matplot++` produces several modules which must be given as an argument to `dependency()`.  The following snippet in my `meson.build` file will build the example `matplotplusplus/examples/line_plot/plot/plot_3.cpp`
```
matplotdep = dependency('Matplot++', 
                        modules: ['Matplot++::nodesoup', 
                                  'Matplot++::cimg', 
                                  'Matplot++::matplot'])
executable('plot_3', sources: ['plot_3.cpp'],
           dependencies: [matplotdep])
```
There is one more possible module `'std::filesystem'` that isn't necessary for `plot_3.cpp`

**Reply by alandefreitas** [reply]
That's interesting. That's probably not what integration would look like if the original build system were meson. "Matplot++" should already link all of that stuff. But this looks good enough for something that's been inferred. Great!
